File: Calc/python/BasicCalc.py
# encoding: utf-8

from PyQt4 import QtCore
from PyQt4.QtCore import pyqtSignal, pyqtSlot

class BasicCalc(QtCore.QObject):
    
    ADD = 0
    SUBS = 1
    MULT = 2
    DIV = 3
    
    operandChanged = pyqtSignal('QString')
    
    def __init__(self, parent = None):
        super(QtCore.QObject, self).__init__(parent)
        self.operand = ''
        self.dataRegister = 0
        self.lastOperation = -1
        self.registerEmpty = True
        
    @pyqtSlot(int)
    def putDigit(self, digit):
        self.operand += str(digit)
        self.operandChanged.emit(self.operand)
        
    @pyqtSlot()
    def clear(self):
        self.operand = ''
        self.registerEmpty = True
        self.operandChanged.emit(self.operand)
        
    @pyqtSlot(int)
    def saveOperation(self, operation):
        if self.registerEmpty:
            self.dataRegister = int(self.operand)
            self.lastOperation = operation
            self.registerEmpty = False
        else:
            self.dataRegister = self.doOperation(self.lastOperation)
            self.lastOperation = operation
            self.operand = str(self.dataRegister)
        self.operandChanged.emit(self.operand)
        self.operand = ''
        
    @pyqtSlot()
    def computeResult(self):
        if self.operand:
            self.dataRegister = self.doOperation(self.lastOperation)
        self.lastOperation = -1
        self.operand = str(self.dataRegister)
        self.operandChanged.emit(self.operand)
        self.operand = ''
        
    def doOperation(self, operation):
        result = self.dataRegister
        
        if operation == self.ADD:
            result = self.dataRegister + int(self.operand)
        elif operation == self.SUBS:
            result = self.dataRegister - int(self.operand)
        elif operation == self.MULT:
            result = self.dataRegister * int(self.operand)
        elif operation == self.DIV:
            result = self.dataRegister / int(self.operand)
            
        return result

File: Calc/ruby/BasicCalc.rb
# encoding: utf-8

require 'Qt'

class BasicCalc < Qt::Object
    ADD = 0
    SUBS = 1
    MULT = 2
    DIV = 3
    
private
    def doOperation(operation)
        result = @dataRegister
        
        if operation == ADD
            result = @dataRegister + @operand.to_i
        elsif operation == SUBS
            result = @dataRegister - @operand.to_i
        elsif operation == MULT
            result = @dataRegister * @operand.to_i
        elsif operation == DIV
            result = @dataRegister / @operand.to_i
        end
        
        return result
    end

public
    slots 'putDigit(int)', 'clear()', 'saveOperation(int)', 'computeResult()'
    signals 'operandChanged(const QString&)'

    def initialize(parent = nil)
        super(parent)
        @operand = ''
        @dataRegister = 0
        @lastOperation = -1
        @registerEmpty = true
    end
    
    def putDigit(digit)
        @operand += digit.to_s
        emit operandChanged(@operand)
    end
    
    def clear()
        @operand = ''
        @registerEmpty = true
        emit operandChanged(@operand)
    end
    
    def saveOperation(operation)
        if @registerEmpty
            @dataRegister = @operand.to_i
            @lastOperation = operation
            @registerEmpty = false
        else
            @dataRegister = doOperation(@lastOperation)
            @lastOperation = operation
            @operand = @dataRegister.to_s
        end
        emit operandChanged(@operand)
        @operand = ''
    end
    
    def computeResult()
        if @operand.empty?
            @dataRegister = doOperation(@lastOperation)
        end
        @lastOperation = -1
        @operand = @dataRegister.to_s
        emit operandChanged(@operand)
        @operand = ''
    end

end
